tools/generate.py

```diff
@@ -1,276 +1,404 @@
+            # some optimization to make connection faster
+            conn.execute("PRAGMA journal_mode=MEMORY;")
+            conn.execute("PRAGMA temp_store=MEMORY;")
+
+            for ngram, hits in ngram_hits.items():
+                ngram = ngram.replace("'", "''")  # escape single quotes
+                conn.execute(
+                    f"""
+                    INSERT INTO {table_name} 
+                      VALUES ('{ngram}', '{lang}', {hits}) 
+                    ON CONFLICT (gram, lang) 
+                      DO UPDATE SET hit = hit + {hits};
+                    """
+                )
+            conn.execute("PRAGMA shrink_memory;")  # reduce memory load
+
+    def _insert_user_scores(self, user_lang_score: dict) -> None:
+        """Insert user language contribution scores into table"""
+
+        print("Inserting users contribution scores")
+        with sqlite3.connect(self._fp) as conn:
+            for (user, lang), hit in user_lang_score.items():
+                conn.execute(
+                    "INSERT INTO users_langs VALUES (?,?,?);",
+                    (user, lang, hit),
+                )
+            conn.execute("PRAGMA shrink_memory;")  # reduce memory load
+
+    @staticmethod
+    def _print_status(
+        line_number: int, ngram_size: int, force: bool = False
+    ) -> None:
+        """Keep track of n-gram counting progress"""
+
+        if line_number % 10000 == 0 or force:
+            msg = (
+                f"\rGenerating ngrams of size {ngram_size} "
+                f"(reading CSV file... {line_number} lines)"
+            )
+            print(msg, end="")
+            sys.stdout.flush()
+
+    @property
+    def path(self) -> Path:
+        """Get the path of the database file"""
+
+        return self._fp
+
+
+class TatodetectDB:
+    """The actual database that is used to detect languages
+    on tatoeba.org
+
+    The Tatodetect algorithm is based on the analysis of the
+    key n-grams found in a Tatoeba monolingual corpus.
 
-    print('Inserting user stats...')
-    for (user,lang),hit in userLangNbrNgram.items():
-        if hit > USR_LANG_LIMIT:
+    This database contains only the key content of the
+    Tatodetect 'raw database'.
+    """
+
+    def __init__(self, db_path: Path) -> None:
+        """
+        Parameters
+        ----------
+        db_path : Path
+            the location of the sqlite database
+        """
+        self._fp = db_path
+
+    def extract_key_content_from(self, raw_db: RawTatodetectDB) -> None:
+        """Import key content from the Tatodetect raw database
+
+        Parameters
+        ----------
+        raw_db : RawTatodetectDB
+            a database containing all Tatoeba n-grams counts and user
+            contribution scores
+        """
+        # the former Tatodetect database is overwritten
+        if self._fp.exists():
+            print(f"Deleting former {self._fp.name} database")
+            self._fp.unlink()
```

Comment:
This code is duplicated in the RawTatodetectDB class. It should be either here or there, but not at both places.